### DISCUSSION TITLE
Meaning of different UID for the same row in the match table

### DISCUSSION THREAD

**Discussion by george-tog** [EnnyvanBeest/UnitMatch#92] [Q&A] [answered]
Thank you for developing such a well-integrated software for Kilosort4 post-processing. 

I was wondering why in some rows of the MatchTable.csv output there are different UID1 and UID2? I think I may have a fundamental confusion about the meaning of UIDs and matches, but I was assuming that a given neuron present in multiple sessions would have the same UID, regardless of the "direction" of matching. What is the sensible way to merge UID1 and UID2, if there is one?
[matched_units.csv](https://github.com/user-attachments/files/17780243/matched_units.csv)

Second, are UID int 1 and UID int 2 the default setting for matching?

**Comment by Sam-Dodgson** [comment]
Hi, good question.
UID1 and UID2 are just ID given to every unit, so each unit has a unique id (as over session there will be multiple units with a KiloSort/phy id 1). So if you look at the UID 2 column you can see this is just ascending number of the UID (there will be gaps in this case if there are no matched units) The UID 1 is then the unique ID of the other unit which it is matched to.

To match units across days we have developed the assign unique ID algorithms, this is the UID Liberal, UID int and UID conservative which   is our way of creating "groups" of matched units which should all correspond to the same Neuron! So in this case when it is say UID Liberal 1 that is a group of different units across days which matched to that units and is likely from the same neuron (if it is the same day UnitMatch is saying the units are very similar or it could be a split unit!).

To summarise (since this matching across days is messy)

- UID is a unique ID number given to every unit to every unit (A unit reference)
- UID Liberal, int and conservative are three different ways of grouping matches together and then given a common unique ID. (A match group reference)

There difference between liberal, intermediate and conservative are explained in the [UnitMatch paper](https://www.nature.com/articles/s41592-024-02440-1#Sec2) in the results section under the sub-heading Tracking Over Many Recordings.

Please let me know if you have any other questions!

**Reply by Electrophysionico** [reply]
I was wondering if there was a way to retrieve the UnitIDs for each recording day using the UniqueID. Lets say the UniqueID is 125 and has been matched for three days, is there a way to find the daily UnitID (for example 45 for day1, 94 for day2 and 145 for day3). The idea would be to investigate the spiking activity of the matched unit through several days.

**Reply by vasilisazhukova** [reply]
@Sam-Dodgson Hello,

in the paper I found a threshold that the default (intermediate) algorithm uses (0,5). Could you please tell which values are used for liberal and concervative algorithms? 

Thank you!

**Reply by Sam-Dodgson** [reply]
So here is how the liberal, intermediate and conservative methods differ:
The code goes through every pair of matched units from highest match probability to lowest match probability and then... 
-     Conservative : adds units which match with EVERY unit in the proposed group
-     Intermediate : adds units which match with EVERY unit in same/adjacent sessions in the proposed group
-     Liberal : adds all units which match with any unit in the proposed group

By default the match probability is 0.5 (if you look at a histogram of the match probabilities, most are very close to 0 or 1) for every method, the different method just have different rules for grouping units together!

**Reply by vasilisazhukova** [reply]
Thank you!

**Comment by cbimbo** [comment, marked as answer]
Hi,
This should be saved in "ID1" for example. So if you find in the MatchTable the rows where UID1 is the one you're looking for, you can check for each session (RecSes1) what is the value for ID1. That should give you the ID of the unit of UID1 tracked across your sessions.
Hope this helps?

**Reply by Electrophysionico** [reply]
Thank you that helped a lot!

**Reply by vasilisazhukova** [reply]
@cbimbo then, the indexes of avg_waveform for resulting waveforms are the indexes of the initial clusters (indexes from kilosort) or ID1?

**Reply by cbimbo** [reply]
Not sure I understand exactly but ID1 and ID2 _are_ the kilosort indices?

**Reply by vasilisazhukova** [reply]
Maybe it is already a different topic, but I am a bit surprised by the results of waveforminfo file, maybe I confused what the indexes mean for each waveform... 

I plotted waveforms from 4 unitmatch runs, one session was the same for all runs, the second one was different for each run. Then, I plotted waveforms from 'avg_waveform' for one specific index [:, 8:9, 0] and [:, 8:9, 1] from 4 resulting files. But intead of having a set of plots with slight differences, I got only two, Why does it happen? 
Are these waveforms the average waveforms from the control group (within the session) or set of pairs from total match or something else?

**Reply by vasilisazhukova** [reply]
@cbimbo yes, I think this is my question. I am attaching a picture. If the cluster 16 from session 1 matches the cluster 10 from session 2, what is UID 8? Is it the indice generated for this match?
![grafik](https://github.com/user-attachments/assets/1ed2434b-9ec4-4244-a195-688340c69bfb)

**Reply by cbimbo** [reply]
UID is the unique ID number given by the tracking algorithm. If here UID1 = UID2 it means that the algorithm thinks they're coming from the same neuron. Its specific value (8) does not mean anything, it's just ~the order it was found by the algo.

Re. your other question, I am not sure I understand. If it sounds like an issue, please open an issue with some more details? Thanks!

**Reply by vasilisazhukova** [reply]
@cbimbo thank you very much! But the ID1 and ID2 are the indices from kilosort, right?

**Reply by cbimbo** [reply]
Yes!
